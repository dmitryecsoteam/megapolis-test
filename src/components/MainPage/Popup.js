import React from 'react';
import PropTypes from 'prop-types';
import ReactPopup from 'reactjs-popup';


class Popup extends React.Component {

    state = {
        title: '',
        error: false
    }

    onTitleChange = ({ target }) => {
        this.setState({
            title: target.value
        });
    }

    onFormSubmit = (e, close) => {
        e.preventDefault();

        if (this.state.title === '') this.setState({
            error: true
        });
        else {
            this.props.addItem(this.state.title);
            close();
        }
    }

    onClose = () => {

        // Clear state each time popup is closed
        this.setState({
            title: '',
            error: false
        });
    }

    render() {

        const overlayStyle = {
            background: 'rgba(242,242,242,0.5)'
        };

        const contentStyle = {
            width: '46.4rem',
            height: '18rem',
            padding: '2.6rem 3.2rem',
            borderRadius: '6px'
        }

        return (
            <ReactPopup
                trigger={<button className='button button--small button--green'>Добавить</button>}
                modal
                closeOnDocumentClick
                overlayStyle={overlayStyle}
                contentStyle={contentStyle}
                onClose={this.onClose}
            >
                {close => (
                    <div className='add-popup__container'>

                        <button
                            className='add-popup__close-btn'
                            onClick={close}
                        >&#10005;</button>

                        <form
                            className='add-popup__form-container'
                            onSubmit={(e) => this.onFormSubmit(e, close)}
                        >

                            <label className='add-popup__label'>Краткое описание</label>
                            <input
                                className='add-popup__input'
                                type="text"
                                value={this.state.title}
                                onChange={this.onTitleChange}
                            />
                            <div className='add-popup__error'>{this.state.error && "Заголовок не может быть пустым"}</div>
                            <button className='button button--small button--green add-popup__button'>Создать</button>

                        </form>


                    </div>
                )
                }
            </ReactPopup>
        );
    }
}

Popup.propTypes = {
    addItem: PropTypes.func.isRequired
}

export default Popup;