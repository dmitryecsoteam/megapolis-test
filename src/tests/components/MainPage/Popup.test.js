import React from 'react';
import { mount } from 'enzyme';
import Popup from '../../../components/MainPage/Popup';

let wrapper;
const addItem = jest.fn();

beforeEach(() => {
    wrapper = mount(<Popup addItem={addItem} />);
    // Показать форму во всплывающем окне
    wrapper.find('button').simulate('click');
});

test('should render Popup component', () => {
    const wrapperClosed = mount(<Popup addItem={addItem} />);
    expect(wrapper).toMatchSnapshot();
    expect(wrapperClosed).toMatchSnapshot();
});

test('should render error message when add item with empty title', () => {
    wrapper.find('form').simulate('submit');
    expect(wrapper.state('error')).toBeTruthy();
    expect(wrapper.find('.add-popup__error')).toMatchSnapshot();
});

test('should add item with non-empty title', () => {
    const value = 'test title';
    wrapper.find('input').simulate('change', { target: { value } });
    expect(wrapper.state('title')).toBe(value);

    wrapper.find('form').simulate('submit');
    expect(addItem).toHaveBeenCalledTimes(1);
    expect(addItem).toHaveBeenCalledWith(value)
})