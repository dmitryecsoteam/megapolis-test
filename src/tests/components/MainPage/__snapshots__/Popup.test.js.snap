// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render Popup component 1`] = `
<Popup
  addItem={[MockFunction]}
>
  <Popup
    arrow={true}
    arrowStyle={Object {}}
    className=""
    closeOnDocumentClick={true}
    closeOnEscape={true}
    contentStyle={
      Object {
        "borderRadius": "6px",
        "height": "18rem",
        "padding": "2.6rem 3.2rem",
        "width": "46.4rem",
      }
    }
    defaultOpen={false}
    disabled={false}
    keepTooltipInside={false}
    lockScroll={false}
    modal={true}
    mouseEnterDelay={100}
    mouseLeaveDelay={100}
    offsetX={0}
    offsetY={0}
    on={
      Array [
        "click",
      ]
    }
    onClose={[Function]}
    onOpen={[Function]}
    open={false}
    overlayStyle={
      Object {
        "background": "rgba(242,242,242,0.5)",
      }
    }
    position="bottom center"
    repositionOnResize={true}
    trigger={
      <button
        className="button button--small button--green"
      >
        Добавить
      </button>
    }
  >
    <button
      className="button button--small button--green"
      key="T"
      onClick={[Function]}
    >
      Добавить
    </button>
    <div
      key="H"
      style={
        Object {
          "left": "0px",
          "position": "absolute",
          "top": "0px",
        }
      }
    />
    <div
      className="popup-overlay "
      key="O"
      onClick={[Function]}
      style={
        Object {
          "background": "rgba(242,242,242,0.5)",
          "bottom": "0",
          "display": "flex",
          "left": "0",
          "position": "fixed",
          "right": "0",
          "top": "0",
          "zIndex": "999",
        }
      }
    >
      <div
        className="popup-content "
        key="C"
        onClick={[Function]}
        style={
          Object {
            "background": "rgb(255, 255, 255)",
            "border": "1px solid rgb(187, 187, 187)",
            "borderRadius": "6px",
            "height": "18rem",
            "margin": "auto",
            "padding": "2.6rem 3.2rem",
            "position": "relative",
            "width": "46.4rem",
          }
        }
      >
        <div
          className="add-popup__container"
        >
          <button
            className="add-popup__close-btn"
            onClick={[Function]}
          >
            ✕
          </button>
          <form
            className="add-popup__form-container"
            onSubmit={[Function]}
          >
            <label
              className="add-popup__label"
            >
              Краткое описание
            </label>
            <input
              className="add-popup__input"
              onChange={[Function]}
              type="text"
              value=""
            />
            <div
              className="add-popup__error"
            />
            <button
              className="button button--small button--green add-popup__button"
            >
              Создать
            </button>
          </form>
        </div>
      </div>
    </div>
  </Popup>
</Popup>
`;

exports[`should render Popup component 2`] = `
<Popup
  addItem={[MockFunction]}
>
  <Popup
    arrow={true}
    arrowStyle={Object {}}
    className=""
    closeOnDocumentClick={true}
    closeOnEscape={true}
    contentStyle={
      Object {
        "borderRadius": "6px",
        "height": "18rem",
        "padding": "2.6rem 3.2rem",
        "width": "46.4rem",
      }
    }
    defaultOpen={false}
    disabled={false}
    keepTooltipInside={false}
    lockScroll={false}
    modal={true}
    mouseEnterDelay={100}
    mouseLeaveDelay={100}
    offsetX={0}
    offsetY={0}
    on={
      Array [
        "click",
      ]
    }
    onClose={[Function]}
    onOpen={[Function]}
    open={false}
    overlayStyle={
      Object {
        "background": "rgba(242,242,242,0.5)",
      }
    }
    position="bottom center"
    repositionOnResize={true}
    trigger={
      <button
        className="button button--small button--green"
      >
        Добавить
      </button>
    }
  >
    <button
      className="button button--small button--green"
      key="T"
      onClick={[Function]}
    >
      Добавить
    </button>
  </Popup>
</Popup>
`;

exports[`should render error message when add item with empty title 1`] = `
<div
  className="add-popup__error"
>
  Заголовок не может быть пустым
</div>
`;
